### ISSUE TITLE
unix "pipe" semantics

### ISSUE THREAD

**Issue by bkreider** [conda/conda-env#50]
Can we change the command-line behavior to be compliant with standard unix tools and read from stdin by default or a specified file?   This lets you use HERE documents and pipes if you want to script this without adding the problems of temporary files.  

```
cat <file> | conda env 
conda env environment.yaml
conda env environment2.yaml
```

**Comment by tswicegood** [comment]
@bkreider I like this idea, but I would do it with the following tweak: in the absence of a `--file`, I think we should perform the following checks:

* Attempt to read `sys.stdin`
* Attempt to read `environment.ya?ml`
* Attempt to read `conda.recipe/environment.ya?ml`

If all of those fail, then let the user know that it's failed.

All that said, I do think we should probably ditch the `--file` in favor of using a standard positional argument.  If we do that, however, we have to continue to support `--file` until we bump to v3.0 as that's a backwards incompatible break.  I'm fine displaying a warning until 3.0, but I don't think its worth a straight BC break just for this feature.

**Comment by bkreider** [comment]
I don't think you can attempt to read sys.stdin.  You block on sys.stdin until there is input, which I think is ok, because explicit is better.

**Comment by tswicegood** [comment]
@bkreider though untested, this is the first thing I found when looking to see how to test stdin: http://stackoverflow.com/a/17735803.  We can't assume that it will always be there because this type of piping may not be available on all platforms.

**Comment by bkreider** [comment]
I don't think you want to assume that tty means there is data to read on stdin.  If you don't want to read from stdin a filename is required.

**Comment by aterrel** [comment]
Personally, I think this request goes against the nature of a conda env.

What you want can be naturally achieved with the conda install command. Conda env, IMHO, is there to build reproducible environments. Having the env.yml file shows exactly how that env was built. Without a file to show how an env was built, there isn't a lot of reason for the env command here.

**Comment by bkreider** [comment]
@aterrel Doesn't that mix policy with implementation?   I'd like clarification that conda-env is limiting itself only to reproducible environments and not setting itself up as the canonical way to manage wholesale environments. There is nothing forcing you to use stdin instead of -f, just like most people use tar with existing directories or archives.   Most people will use a file to define their environments, so it seems like there isn't a downside to making it more flexible.

This forces you to either store a file on disk or deal with the additional complexities of creating temp files no matter your use case.  I could see storing environments in a DB and having reproducible environments that way.  You can even export from one machine and "create" on the other machine by streaming the yaml across without needing to manage actual files.

**Comment by tswicegood** [comment]
@aterrel I think this is an ok thing to add.  I don't think it should be the first-class way of doing it, but I think offering support is an ok thing.

@bkreider what are your thoughts about `cat env.yml | conda env create --file -` and using the `-` option as a way to denote that it should read from stdin?  I've seen that somewhere, git maybe?

**Comment by aterrel** [comment]
@bkreider okay makes sense to me.

**Comment by bkreider** [comment]
Tar uses `-` to represent stdin.  I prefer the standard unix was of defaulting to stdin, but as long as the functionality is there it's down to style.  I'm assuming that "export" will let you do the same thing as tar like this does:

```
tar cpf - /some/important/data | ssh user@destination-machine "tar xpf - -C /some/directory/" 
```

**Comment by dan-blanchard** [comment]
> Tar uses - to represent stdin.

[argparse also supports using `-` to represent stdin.](https://docs.python.org/dev/library/argparse.html#argparse.FileType)

You shouldn't have to even change any code to get `cat env.yml | conda env create --file -` to work.
